src/stats.ts

```diff
@@ -0,0 +1,175 @@
+  private agreements = new Map<string, AgreementInfo>();
+  private providers = new Map<string, ProviderInfo>();
+  private invoices = new Map<string, InvoiceInfo[]>();
+  private payments = new Map<string, PaymentInfo[]>();
+  private times = new Map<string, TimeInfo>();
+  private logger: Logger;
+
+  constructor(
+    private events: EventEmitter<TaskExecutorEventsDict>,
+    options?: StatsServiceOptions,
+  ) {
+    this.logger = options?.logger || defaultLogger("stats");
+  }
+  async run() {
+    this.subscribeGolemEvents();
+    this.subscribeTaskEvents();
+    this.subscribeTimeEvents();
+    this.logger.info("Stats Service has started");
+  }
+
+  async end() {
+    this.unsubscribeAllEvents();
+    this.logger.info("Task Service has stopped");
+  }
+
+  /**
+   * Returns the sum of all costs generated by tasks in the total field, and the sum of all paid invoices in paid field
+   */
+  getAllCosts(): { total: number; paid: number } {
+    const costs = { total: 0, paid: 0 };
+    this.agreements.forEach((agreement) => {
+      const invoices = this.invoices.get(agreement.id);
+      const payments = this.payments.get(agreement.id);
+      costs.total += invoices?.reduce((sum, invoice) => sum + invoice.amount, 0) ?? 0;
+      costs.paid += payments?.reduce((sum, invoice) => sum + invoice.amount, 0) ?? 0;
+    });
+    return costs;
+  }
+
+  /**
+   * Returns an array of agreements and the number of tasks performed on them,
+   * as well as total costs and payment status
+   */
+  getAllCostsSummary() {
+    return [...this.agreements.values()].map((agreement) => {
+      const provider = this.providers.get(agreement.provider.id);
+      const invoices = this.invoices.get(agreement.id);
+      const payments = this.payments.get(agreement.id);
+      const tasks = this.tasksCompleted.get(agreement.id);
+      return {
+        Agreement: agreement.id.substring(0, 10),
+        "Provider Name": provider ? provider.name : "unknown",
+        "Task Computed": tasks?.length ?? 0,
+        Cost: invoices?.reduce((sum, invoice) => sum + invoice.amount, 0),
+        "Payment Status": payments?.length
+          ? payments.length === invoices?.length
+            ? "paid"
+            : "partially-paid"
+          : "unpaid",
+      };
+    });
+  }
+
+  /**
+   * Returns the total computation time (in ms) of all tasks
+   */
+  getComputationTime(): number {
+    return this.times.get("all")?.duration ?? 0;
+  }
+
+  private subscribeTimeEvents() {
+    const startTimeListener = (timestamp: number) => {
+      this.times.set("all", { startTime: timestamp });
+      this.logger.debug("Start time detected", { startTime: timestamp });
+    };
+    this.events.on("ready", startTimeListener);
+    this.listeners.set("ready", startTimeListener);
+
+    const stopTimeListener = (timestamp: number) => {
+      let times = this.times.get("all");
+      if (!times) {
+        times = {};
+        this.times.set("all", times);
+      }
```

Comment:
Why is `times` even a map if it contains at most one key - `all`. Why not just use an object?
```js
private times: TimeInfo = {}
```

Reply:
you're right, an idea for the future, but in this version it's better to simplify it